File: lecture_notes/10 refactoring/5-class.py
import json


class Shift:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def add(self, point, x_key = 'x', y_key = 'y'):
        point[x_key] += self.x
        point[y_key] += self.y
        return point


def parse_point(point):
    return point if (type(point) is dict) else json.loads(point)


polyline = [
    {'x': 0 ,'y': 0},
    {'x': 10, 'y': 10},
    '{"x": 20, "y": 20}',
    {'x': 30, 'y': 30}
]
to_shift = Shift(10, -5)
parsed = list(map(parse_point, polyline))
changed = list(map(to_shift.add, parsed))
for obj in changed:
    print(obj)

File: lecture_notes/10 refactoring/5-class.js
'use strict';

class Shift {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  add(point) {
    point.x += this.x;
    point.y += this.y;
    return point;
  }
}

const parsePoint = (point) => {
  if (typeof point === 'object') point;
  return JSON.parse(point);
};

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const shift = new Shift(10, -5);
const parsed = polyline.map(parsePoint);
const result = parsed.map((point) => shift.add(point));
console.log({ result });
